assets/javascript/apps/pipeline/nodes/widgets.tsx

```diff
@@ -1101,3 +1104,223 @@ function BuiltInToolsWidget(props: WidgetParams) {
+                activateOnTyping: true,
+              }),
+              highlightAutoCompleteVars(autocomplete_vars_list),
+              autocompleteVarTheme(isDarkMode),
+              EditorView.lineWrapping,
+              EditorView.editable.of(true)
+            ]}
+            basicSetup={{
+              lineNumbers: true,
+              tabSize: 2,
+              indentOnInput: true,
+            }}
+          />
+        </div>
+
+        {inputError && <div className="text-red-500">{inputError}</div>}
+      </div>
+      <form method="dialog" className="modal-backdrop">
+        <button>close</button>
+      </form>
+    </dialog>
+  );
+}
+
+function textEditorVarCompletions(autocomplete_vars_list: string[]) {
+  return (context: CompletionContext) => {
+    const word = context.matchBefore(/[a-zA-Z0-9._[\]"]*$/);
+    if (!word || (word.from === word.to && !context.explicit)) return null;
+
+    return {
+      from: word.from,
+      options: autocomplete_vars_list.map((v) => ({
+        label: v,
+        type: "variable",
+        info: `Insert {${v}}`,
+        apply: (view: EditorView, completion: Completion, from: number, to: number) => {
+          const beforeText = view.state.doc.sliceString(from - 1, from);
+          const insertText =
+            beforeText === "{" ? `${completion.label}` : `{${completion.label}}`;
+          view.dispatch({
+            changes: { from, to, insert: insertText },
+          });
+        },
+      })),
+    };
+  };
+}
+// highlight auto complete words. valid - blue, invalid - red
+function highlightAutoCompleteVars(autocomplete_vars_list: string[]) {
+  return ViewPlugin.fromClass(
+    class {
+      decorations: DecorationSet;
+
+      constructor(view: EditorView) {
+        this.decorations = this.buildDecorations(view);
+      }
+
+      update(update: ViewUpdate) {
+        if (update.docChanged || update.viewportChanged) {
+          this.decorations = this.buildDecorations(update.view);
+        }
+      }
+
+      buildDecorations(view: EditorView) {
+        const widgets: any[] = [];
+        const text = view.state.doc.toString();
+        const regex = /\{([a-zA-Z0-9._[\]"]+)\}/g;
+        let match;
+        while ((match = regex.exec(text)) !== null) {
+          const varName = match[1];
+          const from = match.index;
+          const to = from + match[0].length;
+          const isValidVar = autocomplete_vars_list.some(
+            v => varName === v || varName.startsWith(v + ".") || varName.startsWith(v + "[")
+        );
+          const deco = Decoration.mark({
+            class: isValidVar
+              ? "autocomplete-var-valid"
+              : "autocomplete-var-invalid",
+          });
+          widgets.push(deco.range(from, to));
+        }
+        return Decoration.set(widgets);
+      }
+    },
+    {
+      decorations: (v) => v.decorations,
+    }
+  );
+}
+
+const autocompleteVarTheme = (isDarkMode: boolean) =>
+  EditorView.theme({
```

Comment:
Can we change the font to match the font elsewhere in OCS (by default it's using monospace which is OK for code and JSON but not great for prompts):

I think it's this:
```
.cm-editor .cm-content {
    fontFamily: "ui-sans-serif, system-ui, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji'"
}
```

Reply:
Also, in dark mode let's set the background and text color as follows (lower contrast for easier reading)

```
background-color: oklch(25.33% 0.016 252.42)
color: oklch(97.807% 0.029 256.847)
```

(values taken from DaisyUI `--color-base-100` and `--color-base-content`: https://github.com/saadeghi/daisyui/blob/master/packages/daisyui/src/themes/dark.css)